File: Sources/Lib/Config/ResamplerConfigTypes.swift
// Standalone resampler configuration types.

import Foundation

public enum ResamplerType: String, Codable, Equatable, Sendable {
  case synchronous = "Synchronous"
  case apple = "Apple"
  case asyncSinc = "AsyncSinc"
  case asyncPoly = "AsyncPoly"
}

/// Quality settings supported by Apple's AudioConverter.
public enum AppleResamplerQuality: String, Codable, Sendable, CaseIterable, Identifiable, Equatable
{
  case min = "Min"
  case low = "Low"
  case medium = "Medium"
  case high = "High"
  case max = "Max"

  public var id: String { rawValue }
}

/// Algorithm complexity supported by Apple's AudioConverter.
public enum AppleResamplerComplexity: String, Codable, Sendable, CaseIterable, Identifiable,
  Equatable
{
  case linear = "Linear"
  case normal = "Normal"
  case mastering = "Mastering"
  case minimumPhase = "MinimumPhase"

  public var id: String { rawValue }

  public var osType: UInt32 {
    switch self {
    case .linear: return 0x6C69_6E65  // 'line'
    case .normal: return 0x6E6F_726D  // 'norm'
    case .mastering: return 0x6261_7473  // 'bats'
    case .minimumPhase: return 0x6D69_6E70  // 'minp'
    }
  }
}

public struct ResamplerConfig: Codable, Equatable, Sendable, CustomStringConvertible {
  public var type: ResamplerType
  public var profile: String?
  public var interpolation: String?
  public var appleQuality: AppleResamplerQuality?
  public var appleComplexity: AppleResamplerComplexity?
  public var sincLen: Int?
  public var oversamplingFactor: Int?
  public var window: String?
  public var fCutoff: Double?

  enum CodingKeys: String, CodingKey {
    case type
    case profile
    case interpolation
    case appleQuality = "apple_quality"
    case appleComplexity = "apple_complexity"
    case sincLen = "sinc_len"
    case oversamplingFactor = "oversampling_factor"
    case window
    case fCutoff = "f_cutoff"
  }

  public init(
    type: ResamplerType,
    profile: String? = nil,
    interpolation: String? = nil,
    appleQuality: AppleResamplerQuality? = nil,
    appleComplexity: AppleResamplerComplexity? = nil,
    sincLen: Int? = nil,
    oversamplingFactor: Int? = nil,
    window: String? = nil,
    fCutoff: Double? = nil
  ) {
    self.type = type
    self.profile = profile
    self.interpolation = interpolation
    self.appleQuality = appleQuality
    self.appleComplexity = appleComplexity
    self.sincLen = sincLen
    self.oversamplingFactor = oversamplingFactor
    self.window = window
    self.fCutoff = fCutoff
  }

  public var description: String {
    "ResamplerConfig(type: \(type), profile: \(profile ?? "nil"), interpolation: \(interpolation ?? "nil"), sincLen: \(sincLen ?? 0))"
  }
}

File: Sources/CDSP/Config/resampler_config_types.c
#include "resampler_config_types.h"

#include <stdio.h>
#include <string.h>

// Standalone resampler configuration types.

const char* resampler_type_to_string(resampler_type_t type) {
  switch (type) {
    case RESAMPLER_TYPE_SYNCHRONOUS:
      return "Synchronous";
#if defined(ENABLE_COREAUDIO)
    case RESAMPLER_TYPE_APPLE:
      return "Apple";
#endif
    case RESAMPLER_TYPE_ASYNC_SINC:
      return "AsyncSinc";
    case RESAMPLER_TYPE_ASYNC_POLY:
      return "AsyncPoly";
    default:
      return "Synchronous";
  }
}

resampler_type_t resampler_type_from_string(const char* str) {
  if (!str) return RESAMPLER_TYPE_SYNCHRONOUS;
  if (strcmp(str, "Synchronous") == 0) return RESAMPLER_TYPE_SYNCHRONOUS;
#if defined(ENABLE_COREAUDIO)
  if (strcmp(str, "Apple") == 0) return RESAMPLER_TYPE_APPLE;
#endif
  if (strcmp(str, "AsyncSinc") == 0) return RESAMPLER_TYPE_ASYNC_SINC;
  if (strcmp(str, "AsyncPoly") == 0) return RESAMPLER_TYPE_ASYNC_POLY;
  return RESAMPLER_TYPE_SYNCHRONOUS;
}

#if defined(ENABLE_COREAUDIO)
/// Quality settings supported by Apple's AudioConverter.
const char* apple_resampler_quality_to_string(
    apple_resampler_quality_t quality) {
  switch (quality) {
    case APPLE_RESAMPLER_QUALITY_MIN:
      return "Min";
    case APPLE_RESAMPLER_QUALITY_LOW:
      return "Low";
    case APPLE_RESAMPLER_QUALITY_MEDIUM:
      return "Medium";
    case APPLE_RESAMPLER_QUALITY_HIGH:
      return "High";
    case APPLE_RESAMPLER_QUALITY_MAX:
      return "Max";
    default:
      return "Medium";
  }
}

apple_resampler_quality_t apple_resampler_quality_from_string(const char* str) {
  if (!str) return APPLE_RESAMPLER_QUALITY_MEDIUM;
  if (strcmp(str, "Min") == 0) return APPLE_RESAMPLER_QUALITY_MIN;
  if (strcmp(str, "Low") == 0) return APPLE_RESAMPLER_QUALITY_LOW;
  if (strcmp(str, "Medium") == 0) return APPLE_RESAMPLER_QUALITY_MEDIUM;
  if (strcmp(str, "High") == 0) return APPLE_RESAMPLER_QUALITY_HIGH;
  if (strcmp(str, "Max") == 0) return APPLE_RESAMPLER_QUALITY_MAX;
  return APPLE_RESAMPLER_QUALITY_MEDIUM;
}

/// Algorithm complexity supported by Apple's AudioConverter.
const char* apple_resampler_complexity_to_string(
    apple_resampler_complexity_t comp) {
  switch (comp) {
    case APPLE_RESAMPLER_COMPLEXITY_LINEAR:
      return "Linear";
    case APPLE_RESAMPLER_COMPLEXITY_NORMAL:
      return "Normal";
    case APPLE_RESAMPLER_COMPLEXITY_MASTERING:
      return "Mastering";
    case APPLE_RESAMPLER_COMPLEXITY_MINIMUM_PHASE:
      return "MinimumPhase";
    default:
      return "Normal";
  }
}

apple_resampler_complexity_t apple_resampler_complexity_from_string(
    const char* str) {
  if (!str) return APPLE_RESAMPLER_COMPLEXITY_NORMAL;
  if (strcmp(str, "Linear") == 0) return APPLE_RESAMPLER_COMPLEXITY_LINEAR;
  if (strcmp(str, "Normal") == 0) return APPLE_RESAMPLER_COMPLEXITY_NORMAL;
  if (strcmp(str, "Mastering") == 0)
    return APPLE_RESAMPLER_COMPLEXITY_MASTERING;
  if (strcmp(str, "MinimumPhase") == 0)
    return APPLE_RESAMPLER_COMPLEXITY_MINIMUM_PHASE;
  return APPLE_RESAMPLER_COMPLEXITY_NORMAL;
}

uint32_t apple_resampler_complexity_os_type(apple_resampler_complexity_t comp) {
  switch (comp) {
    case APPLE_RESAMPLER_COMPLEXITY_LINEAR:
      return 0x6C696E65;  // 'line'
    case APPLE_RESAMPLER_COMPLEXITY_NORMAL:
      return 0x6E6F726D;  // 'norm'
    case APPLE_RESAMPLER_COMPLEXITY_MASTERING:
      return 0x62617473;  // 'bats'
    case APPLE_RESAMPLER_COMPLEXITY_MINIMUM_PHASE:
      return 0x6D696E70;  // 'minp'
    default:
      return 0x6E6F726D;
  }
}
#endif  // ENABLE_COREAUDIO

const char* resampler_profile_to_string(resampler_profile_t profile) {
  switch (profile) {
    case RESAMPLER_PROFILE_VERY_FAST:
      return "VeryFast";
    case RESAMPLER_PROFILE_FAST:
      return "Fast";
    case RESAMPLER_PROFILE_BALANCED:
      return "Balanced";
    case RESAMPLER_PROFILE_ACCURATE:
      return "Accurate";
    default:
      return "Balanced";
  }
}

resampler_profile_t resampler_profile_from_string(const char* str) {
  if (!str) return RESAMPLER_PROFILE_BALANCED;
  if (strcmp(str, "VeryFast") == 0) return RESAMPLER_PROFILE_VERY_FAST;
  if (strcmp(str, "Fast") == 0) return RESAMPLER_PROFILE_FAST;
  if (strcmp(str, "Balanced") == 0) return RESAMPLER_PROFILE_BALANCED;
  if (strcmp(str, "Accurate") == 0) return RESAMPLER_PROFILE_ACCURATE;
  return RESAMPLER_PROFILE_BALANCED;
}

void resampler_config_init(resampler_config_t* config, resampler_type_t type) {
  if (!config) return;
  memset(config, 0, sizeof(resampler_config_t));
  config->type = type;
}

void resampler_config_description(const resampler_config_t* config,
                                  char* out_buf, size_t buf_len) {
  if (!config || !out_buf || buf_len == 0) return;
  const char* prof = config->has_profile ? config->profile : "nil";
  const char* interp =
      config->has_interpolation ? config->interpolation : "nil";
  int sinc = config->has_sinc_len ? config->sinc_len : 0;
  snprintf(
      out_buf, buf_len,
      "ResamplerConfig(type: %s, profile: %s, interpolation: %s, sincLen: %d)",
      resampler_type_to_string(config->type), prof, interp, sinc);
}
